### PR TITLE
SimpleX Directory Service

### PR DESCRIPTION
**Pull request by epoberezkin**

### CODE REVIEW

File: apps/simplex-directory-service/Options.hs [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+directoryOpts :: FilePath -> FilePath -> Parser DirectoryOpts
+directoryOpts appDir defaultDbFileName = do
+  coreOptions <- coreChatOptsP appDir defaultDbFileName
+  superUsers <-
+    option
+      parseKnownContacts
+      ( long "super-users"
+          <> metavar "SUPER-USERS"
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/Options.hs:28]:
```suggestion
          <> metavar "SUPER_USERS"
```

File: apps/simplex-directory-service/Options.hs [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+          <> help "Comma-separated list of super-users in the format CONTACT_ID:DISPLAY_NAME who will be allowed to manage the directory"
+          <> value []
+      )
+  directoryLog <-
+    strOption
+      ( long "directory-file"
+          <> metavar "DIRECTORY-FILE"
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/Options.hs:35]:
```suggestion
          <> metavar "DIRECTORY_FILE"
```

File: apps/simplex-directory-service/Options.hs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        directoryLog,
+        welcomeMessage
+      }
+
+getDirectoryOpts :: FilePath -> FilePath -> IO DirectoryOpts
+getDirectoryOpts appDir defaultDbFileName =
+  execParser $
+    info
+      (helper <*> versionOption <*> directoryOpts appDir defaultDbFileName)
+      (header versionStr <> fullDesc <> progDesc "Start SimpleX Directory Service with DB_FILE, DIRECTORY-FILE and SUPER-USERS options")
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/Options.hs:58]:
```suggestion
      (header versionStr <> fullDesc <> progDesc "Start SimpleX Directory Service with DB_FILE, DIRECTORY_FILE and SUPER_USERS options")
```

File: apps/simplex-directory-service/src/Directory/Service.hs [hunk 4]
```diff
@@ -0,0 +1,335 @@
...
+      putStrLn $ T.unpack (localDisplayName' ct) <> " connected"
+      sendMessage cc ct $
+        "Welcome to " <> serviceName <> " service!\n\
+        \Send a search string to find groups or */help* to learn how to add groups to directory.\n\n\
+        \For example, send _privacy_ to find groups about privacy."
+
+    deGroupInvitation :: Contact -> GroupInfo -> GroupMemberRole -> GroupMemberRole -> IO ()
+    deGroupInvitation ct g fromMemberRole memberRole =
+      case badInvitation fromMemberRole memberRole of
+        -- TODO check duplicate group name and ask to confirm
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/src/Directory/Service.hs:107]: will both groups be available?

File: apps/simplex-directory-service/src/Directory/Service.hs [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+        GroupInfo {groupId, groupProfile = p} = fromGroup
+        GroupInfo {localDisplayName, groupProfile = p'@GroupProfile {image = image'}} = toGroup
+        groupRef = groupReference toGroup
+        sameProfile
+          GroupProfile {displayName = n, fullName = fn, image = i, description = d}
+          GroupProfile {displayName = n', fullName = fn', image = i', description = d'} =
+            n == n' && fn == fn' && i == i' && d == d'
+        groupLinkAdded gr = do
+          notifyOwner gr $ "Thank you! The group link for " <> groupRef <> " is added to the welcome message.\nYou will be notified once the group is added to the directory - it may take up to 24 hours."
+          let gaId = 1
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/src/Directory/Service.hs:173]: TODO?

File: src/Simplex/Chat.hs [hunk 6]
```diff
@@ -1486,8 +1492,11 @@ processChatCommand = \case
   ListMembers gName -> withUser $ \user -> do
     groupId <- withStore $ \db -> getGroupIdByName db user gName
     processChatCommand $ APIListMembers groupId
-  ListGroups -> withUser $ \user ->
-    CRGroupsList user <$> withStore' (`getUserGroupDetails` user)
+  APIListGroups userId contactId_ search_ -> withUserId userId $ \user ->
```
Comment (spaced4ndy) [inline src/Simplex/Chat.hs:1495]: what's contactId_ for in this api? it's not used in store method

File: apps/simplex-directory-service/src/Directory/Service.hs [hunk 7]
```diff
@@ -0,0 +1,335 @@
...
+      DEContactCommand ct ciId aCmd -> case aCmd of
+        ADC SDRUser cmd -> deUserCommand ct ciId cmd
+        ADC SDRSuperUser cmd -> deSuperUserCommand ct ciId cmd
+  where
+    withSuperUsers action = void . forkIO $ forM_ superUsers $ \KnownContact {contactId} -> action contactId
+    notifySuperUsers s = withSuperUsers $ \contactId -> sendMessage' cc contactId s
+    -- withContact ctId GroupInfo {localDisplayName} err action = do
+    --   getContact cc ctId >>= \case
+    --     Just ct -> action ct
+    --     Nothing -> putStrLn $ T.unpack $ "Error: " <> err <> ", group: " <> localDisplayName <> ", can't find contact ID " <> tshow ctId
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/src/Directory/Service.hs:76]: remove?

File: apps/simplex-directory-service/src/Directory/Service.hs [hunk 8]
```diff
@@ -0,0 +1,335 @@
...
+    notifyOwner GroupReg {dbContactId} = sendMessage' cc dbContactId
+    ctId `isOwner` GroupReg {dbContactId} = ctId == dbContactId
+    withGroupReg GroupInfo {groupId, localDisplayName} err action = do
+      atomically (getGroupReg st groupId) >>= \case
+        Just gr -> action gr
+        Nothing -> putStrLn $ T.unpack $ "Error: " <> err <> ", group: " <> localDisplayName <> ", can't find group registration ID " <> tshow groupId
+    -- withGroupReg' groupId err action = do
+    --   atomically (getGroupReg st groupId) >>= \case
+    --     Just gr -> action gr
+    --     Nothing -> putStrLn $ T.unpack $ "Error: " <> err <> ", can't find group registration ID " <> tshow groupId
```
Comment (spaced4ndy) [inline apps/simplex-directory-service/src/Directory/Service.hs:86]: remove?

File: tests/Bots/DirectoryTests.hs [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+import Directory.Store
+import Simplex.Chat.Bot.KnownContacts
+import Simplex.Chat.Core
+import Simplex.Chat.Options (ChatOpts (..), CoreChatOpts (..))
+import Simplex.Chat.Types (Profile (..), GroupMemberRole (GROwner))
+import System.FilePath ((</>))
+import Test.Hspec
+
+directoryServiceTests :: SpecWith FilePath
+directoryServiceTests = do
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply tests/Bots/DirectoryTests.hs:23]: yes, later

File: apps/simplex-directory-service/src/Directory/Service.hs (hunk 7 above)
Comment (epoberezkin) [reply apps/simplex-directory-service/src/Directory/Service.hs:76]: will clean up later

File: apps/simplex-directory-service/src/Directory/Service.hs [hunk 10]
```diff
@@ -0,0 +1,335 @@
...
+                \Please add it to the group welcome message.\n\
+                \For example, add:"
+              notifyOwner gr $ "Link to join the group " <> T.unpack displayName <> ": " <> B.unpack (strEncode connReqContact)
+            CRChatCmdError _ (ChatError e) -> case e of
+              CEGroupUserRole {} -> notifyOwner gr "Failed creating group link, as service is no longer an admin."
+              CEGroupMemberUserRemoved -> notifyOwner gr "Failed creating group link, as service is removed from the group."
+              CEGroupNotJoined _ -> notifyOwner gr $ unexpectedError "group not joined"
+              CEGroupMemberNotActive -> notifyOwner gr $ unexpectedError "service membership is not active"
+              _ -> notifyOwner gr $ unexpectedError "can't create group link"
+            _ -> notifyOwner gr $ unexpectedError "can't create group link"
```
[the remark this answers is not in the document]
Comment (epoberezkin) [reply apps/simplex-directory-service/src/Directory/Service.hs:137]: it wouldn't even join group unless invited as admin... Possibly, there can be some state management between joining group and adding link (e.g. joining succeeds and adding fails, but I think later.

**Pull request merged**